Fix allocation of link dma channels

ranj063 opened the pull request:
The recommended HDA HW programming sequence for setting the DMA format requires that the link DMA and host DMA channels be coupled before setting the format. This change means that host DMA or link DMA channels be reserved even if only one is used.
    
Statically assigned link DMA channels would mean that all the corresponding host DMA channels will need to be reserved, leaving only a few channels available at run-time. So, the suggestion here is to switch to dynamically assigning both host DMA channels and link DMA channels are run-time.
    
The host DMA channel is assigned when the pcm is opened as before. While choosing the link DMA channel, if the host DMA channel corresponding to the link DMA channel is already taken, the proposed method checks to make sure that the BE is connected to the FE that has been assigned this host DMA channel. For hostless streams, the host DMA channel is marked as opened to ensure it isn't assigned to another stream. After assigning the link DMA channel, a DAI_CONFIG IPC is sent to the FW to set the chosen link DMA channel.

This change should support N:M FE <->BE configurations as well.

The corresponding FW change for this PR is https://github.com/thesofproject/sof/pull/1354

Tested on UP2 so far. Further testing in progress.

wenqingfu commented:
link to issue #874 
replaced PR #882

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -136,7 +77,12 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
+	link_dev = snd_hdac_ext_stream_assign(bus, substream,
+					      HDAC_EXT_STREAM_TYPE_LINK);
+	if (!link_dev)
+		return -EBUSY;
```
doesn't seem quite right:
````c
hdac_ext_link_stream_assign(struct hdac_bus *bus,
				struct snd_pcm_substream *substream)
{
	struct hdac_ext_stream *res = NULL;
	struct hdac_stream *stream = NULL;

	if (!bus->ppcap) {
		dev_err(bus->dev, "stream type not supported\n");
		return NULL;
	}
````
-EINVAL?

ranj063 replied:
Oh yes, should be -EINVAL. will change it .

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -161,11 +163,23 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 	case SNDRV_PCM_TRIGGER_SUSPEND:
 	case SNDRV_PCM_TRIGGER_STOP:
+		if (stream.opened)
+			stream.opened = 0;
```
We directly handle internals of the hdac library but those assignments are only taken into account when there is an explicit call to hdac_ext_xyz(), so it's quite hard to follow the state machine here.

ranj063 replied:
@plbossart understood. I probably need to add some more comments to why and how this is done here. Will address in v2

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -229,8 +241,12 @@ int hda_dsp_stream_trigger(struct snd_sof_dev *sdev,
-	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
 	case SNDRV_PCM_TRIGGER_STOP:
+		/* unlock link DMA */
+		stream->link_locked = 0;
```
same here, you don't actually unlock with this assignment, you just prepare for it to be unlocked, but when that happens is unclear.

ranj063 replied:
@plbossart basically my assumption is this:

1. For a HDA stream, hda_stream_hw_params() would allocate the host DMA channel and hda_link_hw_params() would allocate the link DMA channel. So there will be no issues with coupling/decoupling

2. For an I2S stream for example, hda_stream_hw_params() would allocate the host DMA but there would be no hda_link_hw_params(). So when the START trigger follows, I check if the corresponding link DMA channel is locked or not? If it isnt (which is because this is a I2S playback/capture stream, I set link_locked so that the link DMA channel will not be allocated to the next stream that is started. 

3. Similarly, in the case of hostless streams, there will not be a hda_stream_hw_params() but there will be hda_link_hw_params(). So we would set the stream->opened to true to prevent the host DMA from being assigned to the next stream that is started.

plbossart commented on sound/soc/sof/intel/hda-trace.c:
```diff
@@ -51,6 +51,15 @@ int hda_dsp_trace_init(struct snd_sof_dev *sdev, u32 *stream_tag)
+	 * in decoupled mode, the link DMA channel with the same ID must be
+	 * reserved to avoid it being allocated to another stream.
+	 */
+	hda->dtrace_stream->link_locked = 1;
```
question: this will have no effect when the SOF_HDA option is not selected, e.g. NOCODEC?

ranj063 commented:
> Thanks @ranj063. This PR has the main benefit of yanking all the code I didn't like and the ASoC API abuse, but the parts on the locked/opened flags are a bit unclear, and the ABI change does not seem necessary.

@plbossart I changed the ABI because if the kernel allocates link DMA channels dynamically, it is not compatible with the older version of the FW which require statically allocated channels and vice-versa.

But @lgirdwood commented that he'd like to address the compatibility by having the FW advertise which allocation scheme it uses and the driver can switch to that. This means that we wont be able to remove all the code which abuses the get_channel_map() API. What would be your take on it?

plbossart commented:
> > Thanks @ranj063. This PR has the main benefit of yanking all the code I didn't like and the ASoC API abuse, but the parts on the locked/opened flags are a bit unclear, and the ABI change does not seem necessary.
> 
> @plbossart I changed the ABI because if the kernel allocates link DMA channels dynamically, it is not compatible with the older version of the FW which require statically allocated channels and vice-versa.
> 
> But @lgirdwood commented that he'd like to address the compatibility by having the FW advertise which allocation scheme it uses and the driver can switch to that. This means that we wont be able to remove all the code which abuses the get_channel_map() API. What would be your take on it?

Removing that code solves a long standing issue of abusing the ASoC APIs for HDaudio/iDISP specific stuff, keeping it is not really a goal..

I would err on the side of claiming HDaudio/iDISP never fully worked, and the recommended programming sequence was only communicated recently to the SOF team, and that this is not really a compatibility break but a redesign to deal with how the hardware actually works. We can require all users of Hdaudio/iDISP to update the firmware at this point, and in practice they will update based on new releases already planned. As long as 1.3 includes that release we don't really need to worry.

I hate to break compatibility but here we are enabling a feature.

As discussed I would put a pretend minor ABI change to report that the firmware needs to be upgraded to work around this known incompatibility when HDaudio is used.

ranj063 commented:
@plbossart PR updated now.

lgirdwood commented on include/uapi/sound/sof/abi.h:
```diff
@@ -26,7 +26,7 @@
 /* SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 5
+#define SOF_ABI_MINOR 6
```
MINOR 8

ranj063 replied:
@lgirdwood fixed now

RanderWang commented:
Correct two cases:

````
The sequence of PCM ops upon starting a PCM stream in SOF is as
follows (Showing only the hw_params and trigger ops):

Host hw_params -> Link hw_params -> Host START -> Link START
````
It should be      Link hw_params -> Host hw_params -> Host START -> Link START

````
For HDA streams, the host DMA channel would be assigned during
host hw_params and the corresponding link DMA channel would be
assigned in link hw_params.
````
 Here,  host DMA channel would be assigned in host pcm open (hda_dsp_pcm_open).

RanderWang commented:
I am confused:  The flags: stream->opened and locked are set in start or stop cases, but the dma channel allocation is done at pcm_open for host and hw_param for BE.  How does it work? The host dma and link dma have been allocated before trigger_start, right ?

RanderWang commented on include/sound/sof/dai-intel.h:
```diff
@@ -82,6 +82,7 @@ struct sof_ipc_dai_ssp_params {
 struct sof_ipc_dai_hda_params {
 	struct sof_ipc_hdr hdr;
 	uint32_t link_dma_ch;
+	uint32_t reserved[8];
```
It should be added in sof_ipc_pcm_params for you assign link dma channel in hw_param.
The sof_ipc_dai_hda_params  is for dai_link loading

ranj063 replied:
@RanderWang I suppose i didnt do that good of a job explaining things. Because of our requirement to couple before setting the HDA format, the host DMA stream tag and the link DMA stream tag will always need to be the same. So, I only send one stream tag to the DSP during host pcm hw_params. Because of the way, the link DMA tags are "reserved", the link dma channel assigned during link hw_params will always be the same as host dma stream tag. 

Please give this PR a try and test all kinds of scenarios you think could be broken. That's really what I am doing too. @fredoh9 and myself have been testing a few kinds of concurrent use cases and havent been able to see any issues with our hda generic topology so far.

RanderWang replied:
yes, I just said in FW patch.  two cases:
(1) for two FEs share one BE, the param would be called two times and one of them the tag should be different.
(2) For some streams with only BE like tone generator, there is no host stream

ranj063 replied:
@RanderWang when the hw_params will be the called the second time, we will ignore the channel as it will not be INVALID at that point.

RanderWang replied:
yes, I get it.

RanderWang commented:
And the host stream tag should be different in multi-thread cases.
For example:
(1) pcm1 is opened, and gets host dma channel 1.
(2) pcm2 is opened and gets host dma channel 2,  the hw_params is called for BE of pcm2, so link dma channel 1 is allocated
(3)  the hw_params is called for BE of pcm1, so link dma channel 2 is allocated.

For this sequence, the tag should be different. The key idea is: the open-hw_param-start is not a atomic operation

ranj063 commented:
@RanderWang I'm going to have to challenge you to prove my theory wrong here. Take a look at 
https://github.com/thesofproject/linux/blob/1071743ab1cf1f3e3653088d36967589c916bc5e/sound/soc/soc-pcm.c#L933

pcm_hw_params() is locked before performing fe hw_params and the be hw_params. So what you're describing in the multi-threaded case isnt possible. 

But like I said, this is my theory. Please do the needful to prove me wrong.

RanderWang commented:
this lock only protects BE hw_param, not open function for FE, right ? FE allocates host dma in hda_dsp_pcm_open, which is not protected by this lock

RanderWang commented:
Take it easy. What I asked is discussed with Pierre last year. I spent a least one month.

ranj063 commented:
> Take it easy. What I asked is discussed with Pierre last year. I spent a least one month.

@RanderWang we're talking about discussions that are stale. What you discussed with Pierre last year was before you implemented the programming sequence to couple/decouple while setting the DMA format. Lets not go down a rat hole. Lets stick with the new information and try to make progress.

This change is complicated and is likely not the final solution. But our current solution is severely compromised for concurrent use cases. So lets try to take one step at a time and try to fix the 1:1 FE <-> BE topologies.

ranj063 commented:
> this lock only protects BE hw_param, not open function for FE, right ? FE allocates host dma in hda_dsp_pcm_open, which is not protected by this lock

really? did you take a look at this?

https://github.com/thesofproject/linux/blob/1071743ab1cf1f3e3653088d36967589c916bc5e/sound/soc/soc-pcm.c#L501

RanderWang commented:
> > Take it easy. What I asked is discussed with Pierre last year. I spent a least one month.
> 
> @RanderWang we're talking about discussions that are stale. What you discussed with Pierre last year was before you implemented the programming sequence to couple/decouple while setting the DMA format. Lets not go down a rate hole. Lets stick with the new information and try to make progress.
> 

No, I only debugged couple/decouple case after Feb.  Last year, we worked on link dma allocation.

RanderWang commented:
you don't got my idea.  Each open and hw_param function is protected by lock, but the lock is released after open and acquired when hw_param is invoked ?  So between open and hw_param, there is no lock.

RanderWang commented:
And you could refine our driver,  allocate host dma channel in hw_param for FE, not in open function.
And for 1:1 case, you don't need to set open or lock in trigger function. But you still need to do it for dma trace.

stevyan commented:
@ranj063 This PR would cause an debugfs copying issue, when I tried to copy **_/sys/kernel/debug/sof/dsp_** to other directory, it couldn't be interrupted by any key, even though 'ctrl + c',  and the system also hung. the same operation does't hung system based on daily build version. please help to check it.

**Reproduce Steps**
`sudo su -c 'cp -rf /sys/kernel/debug/sof/dsp /tmp/' `

ranj063 commented:
> @ranj063 This PR would cause an debugfs copying issue, when I tried to copy **_/sys/kernel/debug/sof/dsp_** to other directory, it couldn't be interrupted by any key, even though 'ctrl + c', and the system also hung. the same operation does't hung system based on daily build version. please help to check it.
> 
> **Reproduce Steps**
> `sudo su -c 'cp -rf /sys/kernel/debug/sof/dsp /tmp/' `

@stevyan which platform did this happen on? CML or all of them?

ranj063 commented:
> And you could refine our driver, allocate host dma channel in hw_param for FE, not in open function.
> And for 1:1 case, you don't need to set open or lock in trigger function. But you still need to do it for dma trace.

@RanderWang yes, thats what I am thinking too. Let me change the PR to do host dma stream tag assignment in hw_params. 
But I dont understand why you say that we do not need to set open/locked in trigger. Imagine the case where we start I2S playback with Host DMA 0. Link DMA 0 will not be assigned. Then we start HDMI playback. This will assign host DMA 1 and link DMA 0 for the hDMI playback stream. If we want to avoid this, we have to lock link DMA 0 when the i2s stream was started, isnt it?

RanderWang commented:
> @RanderWang yes, thats what I am thinking too. Let me change the PR to do host dma stream tag assignment in hw_params.
> But I dont understand why you say that we do not need to set open/locked in trigger. Imagine the case where we start I2S playback with Host DMA 0. Link DMA 0 will not be assigned. Then we start HDMI playback. This will assign host DMA 1 and link DMA 0 for the hDMI playback stream. If we want to avoid this, we have to lock link DMA 0 when the i2s stream was started, isnt it?

yes, I agree with you at this case.  One concern is that: do you plan to only support N:1 (N >= 1) for this PR ? As I know, you could archive this without changing ABI or IPC, and your change is ready in FW.
To support 1:N, keyon has offered a solution which needs to at least change ABI, IPC and FW.  I don't require a complete solution in one PR and we may do it in another PR to make it easy to review and test.

Another issue is that: stream should be protected by lock, just like hda_dsp_stream_get. I strongly recommend you improve hda_dsp_stream_get function, and move stream->locked setting to this function for:
(1) it is natural to do it in stream management function and dma trace could also be processed here.

(2) you don't need to protect stream with lock, this function does it.

(3) you could allocate host dma & link dma and reserve corresponding dma channel in hw_params, which is protected by ASoC lock, so there is no multi-thread issue.  If the allocation is done in hw_params but reservation of dma is done in trigger function,  the reserved channel maybe allocated by other pcm in other thread before this thread invokes trigger function. (correct me if I misunderstand)

(4) it is easy to support 1:N case with this change, because all the change should be in hda_dsp_stream_get function.

one basic idea: BE hw_params would be called before FE hw_params by ASoC.

ranj063 commented:
> es, I agree with you at this case. One concern is that: do you plan to only support N:1 (N >= 1) for this PR ? As I know, you could archive this without changing ABI or IPC, and your change is ready in FW.
> To support 1:N, keyon has offered a solution which needs to at least change ABI, IPC and FW. I don't require a complete solution in one PR and we may do it in another PR to make it easy to review and test.

@RanderWang  just discussed this with @keyonjie and we decided to take another route instead of extending the sof_ipc_stream_params structure which would require a major ABI change. 

Instead, we decided that we will send the DAI_CONFIG ipc from hda_link_hw_params(). This DAI config will have no change from when the topology is parsed except for the link DMA channel. The benefit of this approach will be that we may possibly need no change in the FW at all and it might even be backward compatible with the previous solution.

And yes, I will handle all stream locking scenarios in the hw_params in order to avoid the multi-threaded race you pointed out. Let me give this a go today and I will have some updates by tomorrow.

RanderWang commented:
> Instead, we decided that we will send the DAI_CONFIG ipc from hda_link_hw_params(). This DAI config will have no change from when the topology is parsed except for the link DMA channel. The benefit of this approach will be that we may possibly need no change in the FW at all and it might even be backward compatible with the previous solution.
> 
> And yes, I will handle all stream locking scenarios in the hw_params in order to avoid the multi-threaded race you pointed out. Let me give this a go today and I will have some updates by tomorrow.

woo, I want to tell you: I also had this idea yesterday but later I forgot it. I love this idea! Please work as normal and not be so urgent。 Slow work yields fine products (  I got it from google)

xiulipan commented:
@ranj063 
The device hang only on APL_UP_PCM512X.
Also CFL has issue

stevyan commented:
> > @ranj063 This PR would cause an debugfs copying issue, when I tried to copy **_/sys/kernel/debug/sof/dsp_** to other directory, it couldn't be interrupted by any key, even though 'ctrl + c', and the system also hung. the same operation does't hung system based on daily build version. please help to check it.
> > **Reproduce Steps**
> > `sudo su -c 'cp -rf /sys/kernel/debug/sof/dsp /tmp/' `
> 
> @stevyan which platform did this happen on? CML or all of them?

platform info: APL-UP2-pcm512x

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,63 +29,86 @@ struct hda_pipe_params {
+
+	/*
+	 * check if BE is connected to the FE with
+	 * the same stream tag.
```
and this is needed for....???

ranj063 replied:
@plbossart this is needed because of our limitation to surround the DMA format setting with couple and decouple. This happens during the FE hw_params.

So the sequence is:

PCM open -> assign host DMA ch -> link hw_params -> assign link DMA ch -> host hw_params -> couple link and host DMA ch -> set DMA format -> decouple link and host DMA ch

Amd here's the flow: 
1. So when a PCM is opened, the first free host DMA channel is assigned to it. 
2. When we get to link hw_params, we need to find a free link DMA channel but we need to make sure the host DMA channel corresponding to the same link DMA channel is connected to the link. Otherwise, the coupling that happens during host hw_params will break the usecase.

plbossart replied:
I was asking for a short explanation added to this comment...

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,63 +29,86 @@ struct hda_pipe_params {
+				if (hda_check_fes(rtd, stream_dir,
+						  stream->stream_tag)) {
+					res = hstream;
+					goto found;
```
break; ?
why do you need a goto here?

ranj063 replied:
@plbossart oh yes, I had the for loop instead of the function call earlier.  let me remove this.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,63 +29,86 @@ struct hda_pipe_params {
+found:
+	if (res) {
+		if (!res->decoupled)
+			snd_hdac_ext_stream_decouple(bus, res, true);
```
res->decoupled = true?

ranj063 replied:
@plbossart sorry didnt get this. snd_hdac_ext_stream_decouple also sets the res->decouple to true. Is that what you're asking?

plbossart replied:
yes, then add a note that the flag is changed in the hdac library, otherwise it's hard to keep track of the state machine. If you see other cases then same remark.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -122,6 +145,64 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
+		kfree(sof_dai_name);
+	}
+
+	return ret;
```
this is just massively complicated and hard to review.

ranj063 replied:
@plbossart no doubt this is complicated. It took me a while to figure out how to make it work for the N:M case and then implement it while running into some vicious issues with string manipulation. If implementing it was hard, I can only imagine the review to be even harder. 

I probably need to make a design document to explain the methodology to make it easier to review. Let me work on it today.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -147,6 +147,13 @@ static int sof_restore_pipelines(struct snd_sof_dev *sdev)
 
+		/*
+		 * Invalidate link dma channel upon resume.
+		 * This will be assigned during hw_params.
```
why can't we keep the same assignment on resume?

ranj063 replied:
@plbossart because the link DMA channel will change when the link hw_params is done after resume. So its better to remove the stale channel info from the DAI config.

ranj063 commented:
@plbossart addressed your comments now. Also fixed the error seen on CFL.

keyonjie commented on sound/soc/sof/topology.c:
```diff
@@ -2571,9 +2572,7 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
-				struct sof_ipc_dai_config *config,
-				int tx_slot,
-				int rx_slot)
+				struct sof_ipc_dai_config *config)
```
Is it possible to remove those link_hda related here in topology if it is processed in BE hw_params and hda_dai_config IPC sent there?

keyonjie commented on sound/soc/sof/topology.c:
```diff
@@ -2648,10 +2636,6 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
-	u32 tx_num = 0;
-	u32 tx_slot = 0;
-	u32 rx_num = 0;
-	u32 rx_slot = 0;
```
remove all those link_hda related if possible.

keyonjie commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -225,65 +306,44 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
-	}
+	/* free the link DMA channel in the FW */
+	ret = hda_link_config_ipc(hda_stream, dai->name, DMA_CHAN_INVALID,
+				  substream->stream);
```
This is fine to me for short-term solution.
Though here you introduce a new IPC sending, it looks inevitable in current implementation,
```
dai_new()
    dai_config() // get dma channel
        dai_params()
    dai_config(DMA_CHAN_INVALID) // optional, used for HDA dai here to free DMA channel.
dai_free() //free channels for non-HDA dai.
```

The ideal solution for me is something like this:
```
dai_new()
    // dai_config() no needed
        dai_params() // get dma channel
        dai_reset() // free dma channel for all gpdma/link dma.
    // dai_config(DMA_CHAN_INVALID) no needed
dai_free() 
```

ranj063 replied:
@keyonjie i'd like to set the channel in dma_params too but thats going to need some fundamental changes in the firmware in the way the pipelines are set up. Today the params are passed all the way from the host to the DAI. So dai_params is only invoked if host_params is. 
For this we're going to have to split up our pipelines into FE and BE pipelines like we've been discussing off late.

But whats the disadvantage with the current way of using dai_config. Agreed there's a couple of extra IPC's sent just to set the channels but its only one once at at start and once at hw_free. And there's also a request to support multiple DAI configs for each DAI and the ability to switch the config at run-time. This change goes well with that flow I think

keyonjie commented:
@RanderWang Could you please help review to see if there is any concern about @ranj063 latest version?

RanderWang commented:
> @RanderWang Could you please help review to see if there is any concern about @ranj063 latest version?

yes, I am working on it. I have to study this PR and give valuable comments

RanderWang commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -225,65 +312,47 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
+	if (ret < 0)
+		return ret;
+
+	snd_hdac_ext_stream_decouple(bus, link_dev, false);
```
we can't set coupled mode for some cases.  We discussed with windows team and got their usage:
````
we couple DMA engines whenever link DMA engine needs to be reset (so this includes programming FMT and also
audio endpoint being idle and switching to D3 state).
````
For our SOF driver, in link free, we need to check the corresponding FE with the same stream whether it is idle now, .eg. one FE with 2 BEs, one of them will be disable. And host dma free is done in pcm_close which is invoked after link_hw_free, so we need to set dma in coupled mode in FE also.

ranj063 replied:
@RanderWang sorry didnt quite get your point. In the current flow, what I have is decouple while assigning link DMA in link hw_params and couple in hw_free when the link DMA channel is freed. 
Apart from this during host hw_params, before setting the DMA format, your patches already takes care of coupling before and decouple after setting the DMA format. Could you please elaborate on whats incorrect here?

RanderWang replied:
For example, one FE (host dma channel 1) works with two BEs ( link dma channel 1 & 2).  If the BE with link dma 1 is disabled and free by dpcm,  link dam 1 will be coupled with host dma 1 in hda_link_hw_free in your PR.  But this is no correct, because FE is still working with another BE

stevyan commented:
@ranj063 the CI test the base cases failed on the CFL platform due to the topology load failed:
```
[    8.840058 <    0.000004>] sof-audio-pci 0000:00:1f.3: error: tplg component load failed -5
[    8.840123 <    0.000065>] sof-audio-pci 0000:00:1f.3: error: failed to load DSP topology -22
[    8.840175 <    0.000052>] sof-audio-pci 0000:00:1f.3: ASoC: failed to probe component -22
[    8.840261 <    0.000086>] skl_hda_dsp_generic skl_hda_dsp_generic: ASoC: failed to instantiate card -22
[    8.840357 <    0.000096>] skl_hda_dsp_generic: probe of skl_hda_dsp_generic failed with error -22
[    8.913025 <    0.072668>] ax88179_178a 2-7:1.0 enx000ec6a1e622: renamed from eth0
[   17.877758 <    8.964733>] iwlwifi 0000:00:14.3: BIOS contains WGDS but no WRDS
```

**Test recipe**
linux (topic/sof-dev): 35d77594 + PR895
sof (master): 3f1b929b
topology: sof-hda-generic-2ch.tplg

https://sof-ci.01.org/linuxpr/PR895/build1252/build/config-hda-pr895-8.txt

plbossart commented:
Quick write-up that between CI issues and thread comments I don't think this is quite ready for prime time.

ranj063 commented:
@stevyan is the test done with the FW PR 1354?

Jiangxinx commented:
@ranj063 I have tested this patch with thesofproject/sof#1354 on apl-up2-pcm512x platform.
***Test result:***


base | Test result
-- | -- 
Verify firmware presence | Passed
Verify kernel model load probe | Passed
Verify SOF firmware load | Passed
Verify topology binary | Passed
Verify PCM device list | Passed
Check SOF logger | Passed


Analogue Playback Test(hw:0,0) | Test result
-- | -- 
aplay - playback over headset | Passed
stop/resume playback over headset | Passed
Playing -> pause -> s3 -> wakeup -> resume over headset | Passed
repeat 3 times playback over headset | Passed
Pause/Resume playback over headset - 10 times | Passed
Play an audio for 3 minutes | Passed
Suspend & Resume device during audio playback | Passed
Change volume during playback. | Passed


HMDI/DP Playback Test(hw:0, 2) | Test result
-- | -- 
aplay - playback over HDMI | Passed
stop/resume playback over HDMI | Passed
Playing -> pause -> s3 -> wakeup -> resume over HDMI | Passed
repeat 3 times playback over HDMI | Passed
Pause/Resume playback over HDMI | Passed
Play an audio for 3 minutes | Passed
Suspend & Resume device during audio playback | Passed
Change volume during playback. | Passed


Multi-pipeline | Test result
-- | -- 
2 aplay_Analogue + Media Playback | **Failed**
2 aplay_Analogue + HDMI | Passed


General | Test result
-- | -- 
Check the audio channel via speaker-test | Passed
Suspend & Resume device without audio playback or capture | Passed
Check the runtime PM status | Passed
paplay - Play an audio with high-level audio player | Passed
Modules reload | Passed

plbossart commented:
> we need to check FE status when we set link dma to coupled mode.

@RanderWang can you please elaborate. It's not very useful to provide one-liner comments with no references to any code or without explaining why the current solution is incorrect. Thanks

RanderWang commented:
> > we need to check FE status when we set link dma to coupled mode.
> 
> @RanderWang can you please elaborate. It's not very useful to provide one-liner comments with no references to any code or without explaining why the current solution is incorrect. Thanks

@plbossart  I explained it to Ranjani in our discussion two days ago, so I assumed she got it.

```
For example, one FE (host dma channel 1) works with two BEs ( link dma channel 1 & 2).
 If the BE with link dma 1 is disabled and free by dpcm, link dam 1 will be coupled with
host dma 1 in hda_link_hw_free in your PR. But this is no correct, because FE is still
working with another BE
```

ranj063 commented:
> For example, one FE (host dma channel 1) works with two BEs ( link dma channel 1 & 2).
>  If the BE with link dma 1 is disabled and free by dpcm, link dam 1 will be coupled with
> host dma 1 in hda_link_hw_free in your PR. But this is no correct, because FE is still
> working with another BE

@RanderWang sorry I got caught up at OSTS and did not pay attention to your comment. When is this going to happen "If the BE with link DMA 1 is disabled?" Could you please elaborate on when such a scenario could occur? As far as I understand, all the triggers are being passed from the FE to the BE. So I am having a hard time imagining a case where the BE is disabled while is FE continuing on

RanderWang commented:
> @RanderWang sorry I got caught up at OSTS and did not pay attention to your comment. When is this going to happen "If the BE with link DMA 1 is disabled?" Could you please elaborate on when such a scenario could occur? As far as I understand, all the triggers are being passed from the FE to the BE. So I am having a hard time imagining a case where the BE is disabled while is FE continuing on

````
/* Called by DAPM mixer/mux changes to update audio routing between PCMs and
 * any DAI links.
 */
int soc_dpcm_runtime_update(struct snd_soc_card *card)
{
....
}
````

plbossart commented:
@ranj063 can you rebase on top of topic/sof-dev, it'd like to give this a spin for more tests. Thanks

ranj063 commented:
> @ranj063 can you rebase on top of topic/sof-dev, it'd like to give this a spin for more tests. Thanks

@plbossart done. I am going to be testing this over the weekend on the integration branch for 5.0 as well and will update if I find something odd.

ranj063 commented:
> > @RanderWang sorry I got caught up at OSTS and did not pay attention to your comment. When is this going to happen "If the BE with link DMA 1 is disabled?" Could you please elaborate on when such a scenario could occur? As far as I understand, all the triggers are being passed from the FE to the BE. So I am having a hard time imagining a case where the BE is disabled while is FE continuing on
> 
> ```
> /* Called by DAPM mixer/mux changes to update audio routing between PCMs and
>  * any DAI links.
>  */
> int soc_dpcm_runtime_update(struct snd_soc_card *card)
> {
> ....
> }
> ```

@RanderWang @plbossart I just wanted to update a quick discussion I had with Liam last week. The problem with 1:N BE DAI's with HDA is valid but @lgirdwood said that we should call that case as not POR for HDA and move on. With the restriction of having to couple the host and link DMA during link hw_free() makes this cases virtually impossible to fix. I'm not sure if ignoring the coupling during link hw_free is a good idea and I dont have a better way of dealing with it. Ideas welcome. Thanks!

RanderWang commented:
> @RanderWang @plbossart I just wanted to update a quick discussion I had with Liam last week. The problem with 1:N BE DAI's with HDA is valid but @lgirdwood said that we should call that case as not POR for HDA and move on. With the restriction of having to couple the host and link DMA during link hw_free() makes this cases virtually impossible to fix.

here, there is no restriction of having to couple the host and link DMA during link hw_free(). I don't know who declared this. Please read history in the email (Coupled & decoupled mode in HDA) sent by Pierre on Friday.

ranj063 commented:
> here, there is no restriction of having to couple the host and link DMA during link hw_free(). I don't know who declared this. Please read history in the email (Coupled & decoupled mode in HDA) sent by Pierre on Friday.

@RanderWang are you talking about this declaration?

According to Filip's email; "I don’t know the details of cAVS/SST linux driver, so I cannot answer your question. However in the Windows SST driver we couple DMA engines whenever link DMA engine needs to be reset (so this includes programming FMT and **also audio endpoint being idle** and switching to D3 state)."

What does "also audio endpoint being idle" tell you? For me, it definitely sounds like hw_free.

wenqingfu commented:
> > @RanderWang @plbossart I just wanted to update a quick discussion I had with Liam last week. The problem with 1:N BE DAI's with HDA is valid but @lgirdwood said that we should call that case as not POR for HDA and move on. With the restriction of having to couple the host and link DMA during link hw_free() makes this cases virtually impossible to fix.
> 
> here, there is no restriction of having to couple the host and link DMA during link hw_free(). I don't know who declared this. Please read history in the email (Coupled & decoupled mode in HDA) sent by Pierre on Friday.

@RanderWang , I remember @keyonjie was saying the same thing, too. you may want to chat with him. Basically, the safety measure and philosophy is "whenever you don't use them, just couple them back, it cannot be wrong".

keyonjie commented:
Hi @ranj063 @wenqingfu For "also audio endpoint being idle" case, my understand is both the host DMA and link DMA endpoints with the specific index are being idled, then let's set it to couple mode.

I previously suggested @RanderWang to wrap the _set_couple() function and add those checks details inside the wrapped function, then here in hw_free() we can invoke the _set_couple() function blindly.

RanderWang commented:
> > here, there is no restriction of having to couple the host and link DMA during link hw_free(). I don't know who declared this. Please read history in the email (Coupled & decoupled mode in HDA) sent by Pierre on Friday.
> 
> @RanderWang are you talking about this declaration?
> 
> According to Filip's email; "I don’t know the details of cAVS/SST linux driver, so I cannot answer your question. However in the Windows SST driver we couple DMA engines whenever link DMA engine needs to be reset (so this includes programming FMT and **also audio endpoint being idle** and switching to D3 state)."
> 
> What does "also audio endpoint being idle" tell you? For me, it definitely sounds like hw_free.

![endpoint](https://user-images.githubusercontent.com/39945651/58070746-cd55ae00-7bcc-11e9-992d-bba3fd1c8aa6.png)

Do you know what is the audio endpoint ? please check the above picture. there are three endpoints.
Endpoint idle means pcm FE & BE end idle.

@wenqingfu  I have talked with keyon many times and I spent too much time on this thread. Now
I totally understand Pierre.

ranj063 commented:
> @wenqingfu I have talked with keyon many times and I spent too much time on this thread. Now
> I totally understand Pierre.

@RanderWang lets not act pricey here ok? We're all trying to understand and do the right thing here. Some things take time and more clarifications than others.

ranj063 commented:
> Hi @ranj063 @wenqingfu For "also audio endpoint being idle" case, my understand is both the host DMA and link DMA endpoints with the specific index are being idled, then let's set it to couple mode.
> 
> I previously suggested @RanderWang to wrap the _set_couple() function and add those checks details inside the wrapped function, then here in hw_free() we can invoke the _set_couple() function blindly.

@keyonjie I could check if the host DMA channel is free before coupling but what happens if it is not free? When would you ever couple it? 
For example: 1 FE : 2 BE. If you disable one BE, but the FE is still running so you dont couple. When you eventually disable the second BE, you would couple that link DMA channel and host DMA channel but the other one would remain decoupled.

ranj063 commented:
@keyonjie I've updated the PR now. I've added the new patch to take care of the coupling/decoupling issue we discussed. I can squash it if needed. But this will be easier to review. 

@RanderWang Do you mind taking a few minutes of your precious time to see if this addresses your concern? Thanks!

RanderWang commented:
> > @wenqingfu I have talked with keyon many times and I spent too much time on this thread. Now
> > I totally understand Pierre.
> 
> @RanderWang lets not act pricey here ok? We're all trying to understand and do the right thing here. Some things take time and more clarifications than others.

Actually, I repeated a few times, please check the history. I was very patient before.

RanderWang commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -430,6 +438,23 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
+	u32 mask = 0x1 << stream->index;
+
+	/* couple host and link DMA if link DMA channel is idle */
+	if (!link_dev->link_locked)
```
When do you set link_dev->link_locked to zero ?

RanderWang replied:
I get it 。

lyakh commented:
@ranj063 sorry, have you tested S3 on Up^2 with this your patch series and with the corresponding patch for the firmware? I got an immediate failure with them. Reverting them "fixed" the problem.

ranj063 commented:
> @ranj063 sorry, have you tested S3 on Up^2 with this your patch series and with the corresponding patch for the firmware? I got an immediate failure with them. Reverting them "fixed" the problem.

@lyakh thanks for reporting let me test on Up2 today. I have only tested on CML and CFL so far.

ranj063 commented:
> @ranj063 sorry, have you tested S3 on Up^2 with this your patch series and with the corresponding patch for the firmware? I got an immediate failure with them. Reverting them "fixed" the problem.

@lyakh fixed now. I made a stupid mistake of returning an error without checking if it was an error during suspend.

ranj063 commented:
@plbossart I think this PR is almost complete if not fully complete now. Just waiting for test results. Would you mind taking a look?

keyonjie commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -372,7 +372,10 @@ static int hda_link_hw_free(struct snd_pcm_substream *substream,
+	/* couple host and link DMA if host DMA channel is idle */
+	if (!hstream->opened)
+		snd_hdac_ext_stream_decouple(bus, link_dev, false);
+
```
we actually don't need this part as it will be done in line 385 below?
```
snd_hdac_ext_stream_release()
```

ranj063 replied:
@keyonjie good point. removed now

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -134,21 +204,32 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
-	link_dev = snd_soc_dai_get_dma_data(dai, substream);
+	link_dev = hda_link_stream_assign(bus, substream);
+	if (!link_dev)
+		return -EBUSY;
```
why -EBUSY?

ranj063 replied:
@plbossart -EBUSY because hda_link_stream_assign returns a NULL when there are no free link DMA channels available. Is that acceptable?

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -408,6 +408,7 @@ static inline struct hda_bus *sof_to_hbus(struct snd_sof_dev *s)
 }
 
 struct sof_intel_hda_stream {
+	struct snd_sof_dev *sdev;
```
might be worth adding a sdev_to_hda_stream() macro to avoid the repeated container_of in future patches?

ranj063 replied:
@plbossart I am not sure we need this macro. I only use sdev when I have the sof_intel_hda_stream pointer and want to access the sdev. But never the other way around.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -30,62 +30,86 @@ struct hda_pipe_params {
+			continue;
+
+		hda_stream = container_of(hstream, struct sof_intel_hda_stream,
+					  hda_stream);
```
two levels of indirection? what could possibly go wrong here?

ranj063 replied:
@plbossart I dont have a better way of getting the hdac_ext_stream and the sof_intel_hda_stream from the hdac_stream pointer. Do you have a suggestion to avoid this?

plbossart replied:
Using container_of is fine, but you can it more readable with a to_hda_stream(stream); 
same as to_pci_dev().

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -30,62 +30,86 @@ struct hda_pipe_params {
+		 */
+		if (!res->decoupled)
+			snd_hdac_ext_stream_decouple(bus, res, true);
+		spin_lock_irq(&bus->reg_lock);
```
so you take a spinlock here, but not for the ext_stream_decouple() just above? Isn't there a risk if you have a spinlock released between the two parts?

ranj063 replied:
@plbossart snd_hdac_ext_stream_decouple() acquires the spinlock as well.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -30,62 +30,86 @@ struct hda_pipe_params {
+					  hda_stream);
+
+		/* check if available */
+		if (!hstream->link_locked) {
```
what about the else case?

ranj063 replied:
@plbossart the else case just means that we need to move to the next channel

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -217,73 +315,65 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
+					    HDAC_EXT_STREAM_TYPE_LINK);
+
+		/* fallthrough */
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
```
why isn't the TRIGGER_SUSPEND and PAUSE_PUSH not handled in the same way? Both will rely on a .prepare call, no?

ranj063 replied:
@plbossart suspend relies on prepare() upon resume. PAUSE_PUSH doesnt need prepare() when it is just a regular pause_push/release without an S3 in between. I have a separate commit to handle the pause-release case with S3.

plbossart commented on include/sound/sof/dai-intel.h:
```diff
@@ -82,6 +82,7 @@ struct sof_ipc_dai_ssp_params {
 struct sof_ipc_dai_hda_params {
 	struct sof_ipc_hdr hdr;
 	uint32_t link_dma_ch;
+	uint32_t reserved[4];
```
why be greedy, can this be increased to future proof?

Also there's no explanation on why this is needed at all, so will trigger comments upstream.

lyakh replied:
This really doesn't belong here...

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -30,62 +30,86 @@ struct hda_pipe_params {
+}
+
+static struct hdac_ext_stream *
+hda_link_stream_assign(struct hdac_bus *bus,
```
I think it's usual to leave at least one TAB in such splits?

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -122,6 +146,52 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
+			continue;
+
+		if (!strcmp(dai_name, sof_dai->cpu_dai_name)) {
+			if (dir == sof_dai->comp_dai.direction) {
```
I still think it would be better to merge these two "if"s into one

xiulipan commented:
Manually run with latest https://github.com/thesofproject/sof/pull/1354 binary

plbossart commented:
@ranj063 can you update this PR? (no ABI change and @lyakh + my comments addressed) and we should merge. we've hacked this to death.

ranj063 commented:
@plbossart @lyakh updated to address the comments now.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -217,73 +309,64 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 	default:
 		return -EINVAL;
 	}
+
```
remove newline.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -16,6 +16,7 @@
 #include "ops.h"
 
 #define COMP_ID_UNASSIGNED		0xffffffff
+
```
different patch

ranj063 replied:
@plbossat did you mean to separate out the topology parts into a separate patch?

plbossart replied:
no, move the newline somewhere else

ranj063 replied:
@plbossart ahh ok. I've removed it. I dont know how it sneaked in.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -134,21 +201,31 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
 	struct hda_pipe_params p_params = {0};
 	struct hdac_ext_link *link;
-	int stream_tag;
+	int stream_tag, ret;
```
separate line

plbossart commented:
Alright let's merge and test. Thanks @ranj063

plbossart commented:
@xiulipan @wenqingfu the CI is completely in the weeds. The build fail is in the make bindebpkg phase.

plbossart commented:
The compilation works fine on my machine so I will go ahead and merge.

xiulipan commented:
@plbossart @ranj063 
The build fail is because there is a warnging and we treat warning as error in Travis.

```
sound/soc/sof/intel/hda-dsp.c: In function ‘hda_dsp_set_hw_params_upon_resume’:
sound/soc/sof/intel/hda-dsp.c:469:6: error: variable ‘stream_tag’ set but not used [-Werror=unused-but-set-variable]
  int stream_tag;
      ^~~~~~~~~~
cc1: all warnings being treated as errors
```

plbossart commented:
On 5/29/19 4:06 AM, xiulipan wrote:
> @plbossart <https://github.com/plbossart> @ranj063 
> <https://github.com/ranj063>
> The build fail is because there is a warnging and we treat warning as 
> error in Travis.
> 
> |sound/soc/sof/intel/hda-dsp.c: In function 
> ‘hda_dsp_set_hw_params_upon_resume’: 
> sound/soc/sof/intel/hda-dsp.c:469:6: error: variable ‘stream_tag’ set 
> but not used [-Werror=unused-but-set-variable] int stream_tag; 
> ^~~~~~~~~~ cc1: all warnings being treated as errors |

Ah yes. With CI completely gone yesterday we didn't see this. Will send 
a PR to quickly fix this.
